File: python/tflite_micro/signal/ops/fft_ops.py
# Copyright 2022 The TensorFlow Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================
"""Use FFT ops in python."""

import math
import tensorflow as tf
from tflite_micro.python.tflite_micro.signal.utils import util

gen_fft_ops = util.load_custom_op('fft_ops.so')

_MIN_FFT_LENGTH = 64
_MAX_FFT_LENGTH = 2048


def get_pow2_fft_length(input_length):
  """Returns the smallest suuported power of 2 FFT length larger than or equal

  to the input_length.

  Only returns FFT lengths that are powers of 2 within the range
  [_MIN_FFT_LENGTH, _MAX_FFT_LENGTH].

  Args:
    input_length: Length of input time domain signal.

  Returns:
    A pair: the smallest length and its log2 (number of bits)

  Raises:
    ValueError: The FFT length needed is not supported
  """
  fft_bits = math.ceil(math.log2(input_length))
  fft_length = pow(2, fft_bits)
  if not _MIN_FFT_LENGTH <= fft_length <= _MAX_FFT_LENGTH:
    raise ValueError("Invalid fft_length. Must be between %d and %d." %
                     (_MIN_FFT_LENGTH, _MAX_FFT_LENGTH))
  return fft_length, fft_bits


def _fft_wrapper(fft_fn, default_name):
  """Wrapper around gen_fft_ops.*rfft*."""

  def _fft(input_tensor, fft_length, name=default_name):
    if not ((_MIN_FFT_LENGTH <= fft_length <= _MAX_FFT_LENGTH) and
            (fft_length % 2 == 0)):
      raise ValueError(
          "Invalid fft_length. Must be an even number between %d and %d." %
          (_MIN_FFT_LENGTH, _MAX_FFT_LENGTH))
    with tf.name_scope(name) as name:
      input_tensor = tf.convert_to_tensor(input_tensor)
      return fft_fn(input_tensor, fft_length=fft_length, name=name)

  return _fft


def _fft_auto_scale_wrapper(fft_auto_scale_fn, default_name):
  """Wrapper around gen_fft_ops.fft_auto_scale*."""

  def _fft_auto_scale(input_tensor, name=default_name):
    with tf.name_scope(name) as name:
      input_tensor = tf.convert_to_tensor(input_tensor, dtype=tf.int16)
      dim_list = input_tensor.shape.as_list()
      if len(dim_list) != 1:
        raise ValueError("Input tensor must have a rank of 1")
      return fft_auto_scale_fn(input_tensor, name=name)

  return _fft_auto_scale


rfft = _fft_wrapper(gen_fft_ops.signal_rfft, "signal_rfft")
irfft = _fft_wrapper(gen_fft_ops.signal_irfft, "signal_irfft")
fft_auto_scale = _fft_auto_scale_wrapper(gen_fft_ops.signal_fft_auto_scale,
                                         "signal_fft_auto_scale")
tf.no_gradient("signal_rfft")
tf.no_gradient("signal_irfft")
tf.no_gradient("signal_fft_auto_scale")

File: signal/tensorflow_core/ops/fft_ops.cc
/* Copyright 2023 The TensorFlow Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

#include "tensorflow/core/framework/op.h"
#include "tensorflow/core/framework/shape_inference.h"

using tensorflow::shape_inference::InferenceContext;
using tensorflow::shape_inference::ShapeHandle;

namespace tensorflow {
namespace signal {

absl::Status RfftShape(InferenceContext* c) {
  ShapeHandle out;
  int fft_length;
  TF_RETURN_IF_ERROR(c->GetAttr<int>("fft_length", &fft_length));
  TF_RETURN_IF_ERROR(c->WithRankAtLeast(c->input(0), 1, &out));
  auto dim = ((fft_length / 2) + 1) * 2;  // * 2 for complex
  TF_RETURN_IF_ERROR(c->ReplaceDim(out, -1, c->MakeDim(dim), &out));
  c->set_output(0, out);
  return absl::OkStatus();
}

absl::Status IrfftShape(InferenceContext* c) {
  ShapeHandle out;
  int fft_length;
  TF_RETURN_IF_ERROR(c->GetAttr<int>("fft_length", &fft_length));
  TF_RETURN_IF_ERROR(c->WithRankAtLeast(c->input(0), 1, &out));
  TF_RETURN_IF_ERROR(c->ReplaceDim(out, -1, c->MakeDim(fft_length), &out));
  c->set_output(0, out);
  return absl::OkStatus();
}

// TODO(b/286250473): change back name after name clash resolved
REGISTER_OP("SignalRfft")
    .Attr("T: {float, int16, int32}")
    .Attr("fft_length: int >= 2")
    .Input("input: T")
    .Output("output: T")
    .SetShapeFn(RfftShape)
    .Doc(R"doc(
Computes the 1-dimensional discrete Fourier transform of a real-valued signal
over the inner-most dimension of input. Since the DFT of a real signal is
Hermitian-symmetric, RFFT only returns the fft_length / 2 + 1 unique complex
components of the FFT: the zero-frequency term, followed by the fft_length / 2
positive-frequency terms. Along the axis RFFT is computed on, if fft_length is
larger than the corresponding dimension of input, the dimension is padded with
zeros.

input: A Tensor. Must be one of the following types: float32, int16, int32
output: A tensor containing ((fft_length / 2) + 1) complex spectral
        components along its innermost dimension.
        Since there's no TF integer complex type, the array is represented using
        ((fft_length / 2) + 1) * 2 real elements.
        For integer input (int16, int32), the output is scaled by 1 / fft_length
        relative to the theoretical DFT, to avoid overflowing.
        For floating point (float32) input, the output isn't scaled.

fft_length: The length of the FFT operation. An input signal that's shorter
            will be zero padded to fft_length.
)doc");

// TODO(b/286250473): change back name after name clash resolved
REGISTER_OP("SignalIrfft")
    .Attr("T: {float, int16, int32}")
    .Attr("fft_length: int >= 2")
    .Input("input: T")
    .Output("output: T")
    .SetShapeFn(IrfftShape)
    .Doc(R"doc(
Computes the inverse 1-dimensional discrete Fourier transform of a real-valued
signal over the inner-most dimension of input.

The inner-most dimension of input is assumed to be the result of RFFT:
the fft_length / 2 + 1 unique components of the DFT of a real-valued signal.
fft_length must be provided.

input: A tensor containing ((fft_length / 2) + 1) complex spectral
       components along its innermost dimension.
       Since there's no TF integer complex type, the array is represented using
       ((fft_length / 2) + 1) * 2 real elements.
output: A tensor containing fft_length time domain elements along its innermost
        dimension.

fft_length: The length of the IFFT operation.
)doc");

// TODO(b/286250473): change back name after name clash resolved
REGISTER_OP("SignalFftAutoScale")
    .Input("input: int16")
    .Output("output: int16")
    .Output("scale_bits: int32")
    .SetShapeFn([](shape_inference::InferenceContext* c) {
      ShapeHandle out;
      TF_RETURN_IF_ERROR(c->WithRank(c->input(0), 1, &out));
      c->set_output(0, out);
      c->set_output(1, c->Scalar());
      return absl::OkStatus();
    })
    .Doc(R"doc(
Shifts the input left until the amplitude is maximized without clipping. Returns
the amount of left shift for compensation later. This op can be used to maximize
precision of integer FFT implementations, especially 16-bit.

input: A 1-D time domain signal.
output: A 1-D time domain signal after auto scaling.
scale_bits: Scalar. The number of left shifts applied to the input signal.
)doc");

}  // namespace signal
}  // namespace tensorflow
